### PR TITLE
Create orders as 'auto-draft' by default in admin

### LINKED ISSUE #36749: With the HPOS enabled, opening the "add order" page without saving anything still creates an order
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

With the HPOS disabled, opening the Add Order page creates a post draft, which is not automatically saved as an order. If one closes or reloads the page, without explicitly saving the order, the orders list page doesn't show any new orders.

With the HPOS enabled, a new order is automatically created and persisted as soon as the Add Order page is loaded. If the newly created order is abandoned (i.e. never saved), it still appears in the orders list page. Here's an example of the result, which occurred while I was testing a feature on the Add Order page and I reloaded it, without saving anything, several times:

image

### Expected behavior

Opening the Add Order page should not automatically persist an order. Reloading the Add Order page should not create new orders that appear in the Order List page.

### Actual behavior

A new order is created every time the Add Order page is loaded, even if the order is not explicitly saved. The Orderst List page shows all these new orders, which are empty and have the `pending` status

### Steps to reproduce

**Method 1**
1. Go to WooCommerce > Orders.
2. Click on Add Order.
3. **Don't save the order** and just reload the page a couple of times

**Method 2**
1. Go to WooCommerce > Orders.
2. Open new tabs by right-clicking on "Add Order".

### WordPress Environment

`
### WordPress Environment ###

WordPress address (URL): http://localhost/wordpress_wc80
Site address (URL): http://localhost/wordpress_wc80
WC Version: 7.4.0
REST API Version: ✔ 7.4.0
WC Blocks Version: ✔ 9.4.2
Action Scheduler Version: ✔ 3.5.4
Log Directory Writable: ✔
WP Version: 6.0.3
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.35 (Win64) OpenSSL/1.1.1q PHP/8.0.22
PHP Version: 8.0.22
PHP Post Max Size: 2 GB
PHP Time Limit: 36000
PHP Max Input Vars: 1000
cURL Version: 7.76.1
OpenSSL/1.1.1q

SUHOSIN Installed: –
MySQL Version: 5.7.24
Max Upload Size: 2 GB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ❌ Your server does not have the SoapClient class enabled - some gateway plugins which use SOAP may not work as expected.
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 7.4.0
WC Database Prefix: wp_
Total Database Size: 52.46MB
Database Data Size: 28.54MB
Database Index Size: 23.92MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.08MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 1.52MB + Index: 0.78MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.08MB + Index: 0.13MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.06MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.27MB + Index: 0.36MB + Engine InnoDB
wp_e_submissions: Data: 0.02MB + Index: 0.27MB + Engine InnoDB
wp_e_submissions_actions_log: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_e_submissions_values: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_groups_capability: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_groups_group: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_groups_group_capability: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_groups_user_capability: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_groups_user_group: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 6.44MB + Index: 1.17MB + Engine InnoDB
wp_postmeta: Data: 4.52MB + Index: 6.55MB + Engine InnoDB
wp_posts: Data: 1.52MB + Index: 0.22MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.09MB + Index: 0.17MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.08MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.06MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.11MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 1.52MB + Index: 3.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.09MB + Index: 0.09MB + Engine InnoDB
wp_wc_order_bundle_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_order_composite_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.08MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.08MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.06MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.05MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_bundled_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_bundled_items: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_exported_csv_items: Data: 2.52MB + Index: 0.14MB + Engine InnoDB
wp_woocommerce_gpf_google_taxonomy: Data: 1.52MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_mnm_bundled_items: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_pickup_locations_geodata: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_woocommerce_pickup_locations_relationships: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_shipping_table_rates: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wss_sell_messages: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wss_service_notifications: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 26
def_reg_fields: 10
enr_subsc_plan: 1
nav_menu_item: 14
page: 36
post: 5
product: 84
product_variation: 166
revision: 61
shop_coupon: 11
shop_order: 291
shop_order_placehold: 3
shop_order_refund: 3
shop_subscription: 2
shopping-cart: 2
user_request: 1
wc_booking: 12
wp_global_styles: 1
wp_navigation: 1

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (2) ###

Loco Translate: by Tim Whitlock – 2.6.3
WooCommerce: by Automattic – 7.4.0-beta.2

Post SMTP: by Post SMTP – 2.3.2
WooCommerce Beta Tester: by WooCommerce – 2.1.0

### Settings ###

API Enabled: ✔
Force SSL: –
Currency: EUR (€)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: 
external (external)
grouped (grouped)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: –
Order datastore: WC_Order_Data_Store_CPT
HPOS data sync enabled: ✔

### WC Pages ###

Shop base: #4 - /
Cart: #5 - /cart/
Checkout: #6 - /checkout/
My account: #7 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront Child
Version: 1.0.0.180829
Author URL: http://woocommerce.com
Child Theme: ✔
Parent Theme Name: Storefront
Parent Theme Version: 4.1.3 – 4.2.0 is available
Parent Theme Author URL: https://woocommerce.com/
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
coupons
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
multichannel-marketing
mobile-app-banner
navigation
new-product-management-experience
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: minified-js
product-variation-management
settings

Daily Cron: ✔ Next scheduled: 2023-02-04 11:51:20 +01:00
Options: ✔
Notes: 112
Onboarding: completed

### Action Scheduler ###

Complete: 126
Oldest: 2023-01-27 17:36:58 +0000
Newest: 2023-02-03 12:00:50 +0000

Pending: 12
Oldest: 2023-02-03 12:41:34 +0000
Newest: 2023-02-22 14:07:49 +0000

Failed: 27
Oldest: 2021-12-15 13:13:11 +0000
Newest: 2023-02-03 12:07:15 +0000

### Status report information ###

Generated at: 2023-02-03 13:40:37 +01:00
`

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by jorgeatorres**
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

When CPT is authoritative, adding new orders in the admin results in the underlying post being created as 'auto-draft' until explicitly saved. This keeps the orders list tidier and allows WP to automatically clear up those orders after a while during a scheduled deletion.

This PR adds the same functionality to HPOS authoritative.

Closes #36749.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Test that new orders are created as 'auto-draft' by default
1. Make HPOS authoritative: to do so, go to WooCommerce > Settings > Advanced > Features and make sure "Enable the high performance order storage feature." is checked off. Similarly, make sure "Use the WooCommerce orders tables" is selected in WooCommerce > Settings > Advanced > Custom data stores.
2. Go to WooCommerce > Orders.
3. Click "Add order" at the top and take note of the order ID in the details metabox. Do not save the order.
5. Return to WooCommerce > Orders.
6. Confirm that the new order is not listed when the "All" filter is active.
7. Confirm that there's an "Auto Draft" status in the top filter on this screen.
8. Click "Auto Draft" and make sure the order created in step 3 is listed here.
9. Repeat step 3 but now do save the order, after having chosen a status in the "Status" dropdown.
10. Confirm that the order is listed under its respective status in WooCommerce > Orders.

#### Test that auto-draft orders are deleted after a while
We're using WP-CLI for this as we need to "fake" an old order.

1. Make sure HPOS is authoritative as above.
2. Use `wp shell` in your test site's WP dir, to run the following set of commands/code:
   ```
   > $order = new \WC_Order();
   > $order->set_date_created( '2023-04-01 00:00:00' );
   > $order->set_status( 'auto-draft' );
   > $order->save();
   ```
   Take note of the order ID that gets printed when the last command is executed (it should look something like `=> int(8031)`.
3. Go to WooCommerce > Orders > Auto Draft.
4. Confirm the order with the ID from step 2 is listed.
5. Manually run the "wp_scheduled_auto_draft_delete" scheduled hook. To do so, you can invoke `do_action( 'wp_scheduled_auto_draft_delete' );` in a `wp shell`, or you can use a plugin such as [WP Crontrol](https://wordpress.org/plugins/wp-crontrol/).
   If using the latter, go to Tools > Cron Events. Choose "WordPress core events" and find the row with hook `wp_scheduled_auto_draft_delete`. Click "Run now".
6. Go to WooCommerce > Orders > Auto Draft and confirm that the order from step 2 is no longer listed.

### CODE REVIEW

File: plugins/woocommerce/includes/abstracts/abstract-wc-order.php [hunk 1]
```diff
@@ -638,7 +638,7 @@ public function set_status( $new_status ) {
 			}
 
 			// If the old status is set but unknown (e.g. draft) assume its pending for action usage.
-			if ( $old_status && ! in_array( 'wc-' . $old_status, $this->get_valid_statuses(), true ) && ! in_array( $old_status, $status_exceptions, true ) ) {
+			if ( $old_status && ( ( ! in_array( 'wc-' . $old_status, $this->get_valid_statuses(), true ) && ! in_array( $old_status, $status_exceptions, true ) ) || 'auto-draft' === $old_status ) ) {
```
Comment (rrennick) [inline plugins/woocommerce/includes/abstracts/abstract-wc-order.php]: Adding the new condition first will allow skipping the `in_array` scans when it's an `auto-draft`

```suggestion
			if ( $old_status && ( 'auto-draft' === $old_status || ( ! in_array( 'wc-' . $old_status, $this->get_valid_statuses(), true ) && ! in_array( $old_status, $status_exceptions, true ) ) ) {
```

File: plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php [hunk 2]
```diff
@@ -79,6 +79,8 @@ public function __construct() {
 		self::add_action( 'deleted_post', array( $this, 'handle_deleted_post' ), 10, 2 );
 		self::add_action( 'woocommerce_new_order', array( $this, 'handle_updated_order' ), 100 );
 		self::add_action( 'woocommerce_update_order', array( $this, 'handle_updated_order' ), 100 );
+		self::add_action( 'wp_scheduled_auto_draft_delete', array( $this, 'delete_auto_draft_orders' ) );
```
Comment (rrennick) [inline plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php]: Optional: Should we hook this at a lower priority than the WP core auto-draft cleaner?

**Review by rrennick — COMMENTED** [review]
@jorgeatorres This tested great. I just had the two questions/comments. Once we discuss those this is good to merge.

File: plugins/woocommerce/includes/abstracts/abstract-wc-order.php (hunk 1 above)
Comment (jorgeatorres) [reply plugins/woocommerce/includes/abstracts/abstract-wc-order.php]: Makes sense. Implemented in 3c27123aa2cf83c8d579846e5ebd3bb733c368aa.

File: plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php (hunk 2 above)
Comment (jorgeatorres) [reply plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php]: Implemented in 3c27123aa2cf83c8d579846e5ebd3bb733c368aa. It makes sense, that way posts corresponding to orders are deleted too making WP's job a bit easier.

**Comment by jorgeatorres** [comment]
Hi @rrennick!

Thanks for the review. I've implemented the suggested changes.

**Review by rrennick — APPROVED** [review]

**Pull request merged**